src/java.base/share/classes/java/text/ListFormat.java

```diff
@@ -0,0 +1,571 @@
+     * If an error occurs, then the index of {@code parsePos} is not
+     * changed, the error index of {@code parsePos} is set to the index of
+     * the character where the error occurred, and null is returned.
+     * See the {@link #parse(String)} method for more information
+     * on list parsing.
+     *
+     * @param source A {@code String}, part of which should be parsed.
+     * @param parsePos A {@code ParsePosition} object with index and error
+     *            index information as described above.
+     * @return A {@code List} of {@code String} parsed from the string. In case of
+     *         error, returns null.
+     * @throws NullPointerException if {@code source} or {@code parsePos} is null.
+     */
+    @Override
+    public Object parseObject(String source, ParsePosition parsePos) {
+        Objects.requireNonNull(source);
+        Objects.requireNonNull(parsePos);
+        var sm = startPattern.matcher(source);
+        var em = endPattern.matcher(source);
+        Object parsed = null;
+        if (sm.find(parsePos.index) && em.find(parsePos.index)) {
+            // get em to the last
+            var c = em.start();
+            while (em.find()) {
+                c = em.start();
+            }
+            em.find(c);
+            var startEnd = sm.end();
+            var endStart = em.start();
+            if (startEnd <= endStart) {
+                var mid = source.substring(startEnd, endStart);
+                var count = mid.split(middleBetween).length + 2;
+                parsed = new MessageFormat(createMessageFormatString(count), locale).parseObject(source, parsePos);
+            }
+        }
+
+        if (parsed == null) {
+            // now try exact number patterns
+            parsed = new MessageFormat(patterns[TWO], locale).parseObject(source, parsePos);
+            if (parsed == null && !patterns[THREE].isEmpty()) {
+                parsed = new MessageFormat(patterns[THREE], locale).parseObject(source, parsePos);
+            }
+        }
+
+        // return the entire source from parsePos if still no match
+        if (parsed == null) {
+            parsed = new String[]{source.substring(parsePos.index)};
+            parsePos.setIndex(source.length());
+        }
+
+        if (parsed instanceof Object[] objs) {
+            parsePos.setErrorIndex(-1);
+            return Arrays.asList(objs);
+        }
+        throw new InternalError("MessageFormat.parseObject() should return Object[]");
+    }
+
+    @Override
+    public AttributedCharacterIterator formatToCharacterIterator(Object arguments) {
+        if (arguments instanceof List<?> objs) {
+            var a = objs.toArray(new Object[0]);
+            return generateMessageFormat(a).formatToCharacterIterator(a);
+        } else {
+            throw new IllegalArgumentException("The arguments should be an Object list");
+        }
+    }
+
+    /**
+     * {@inheritDoc}
```

Comment:
The javadoc from Object.equals doesn't give any information about what constituted equals for a ListFormat.
Namely, that the locale and patterns are equal.

Reply:
Will add those ListFormat specific considerations.